### ISSUE TITLE
[Enhancement]: Improve provider documentation

### ISSUE THREAD

**Issue by swade1987** [fluxcd/terraform-provider-flux#652]
### Description

I want to improve the provider documentation on the [Terraform registry website](https://registry.terraform.io/providers/fluxcd/flux/latest/docs). 
Please add comments on what you would like covered.

### References

⚠️ Warning ⚠️ 
Please note that the new documentation (once released) will be the current contents of the `docs` directory [here](https://github.com/fluxcd/terraform-provider-flux/tree/main/docs). 
Please use this directory's contents when making suggestions.
I have deleted all guides and have created a suite of new examples available [here](https://github.com/fluxcd/terraform-provider-flux/tree/main/examples)

### Would you like to implement a fix?

Yes

**Comment by bwestover** [comment]
Thanks for asking for input. Apologies if this is not the right place for this :blush: 

When looking at the current example in the `docs` directory I noticed that they all seem create a new Git repository. This is great for a working example :+1: (it also creates a kind cluster)

I have a mono-repo with several clusters already set up with Flux (using the CLI). Now I'm looking to use `flux_bootstrap_git` to add a new cluster to this existing repository (in a new directory I've set up for this cluster). An example on how to use the provider to bootstrap a new cluster with an *existing* repository would be helpful for me, and hopefully others as well.

**Comment by swade1987** [comment]
@bwestover 

Firstly, apologies for the late reply.

Are you looking to use `flux_bootstrap_git` with a cluster that already has flux v2 installed into it? If so it is recommended not to use `flux_bootstrap_git` as the cluster is already bootstrapped. Instead, it's recommended to use the upstream helm chart, see the example of how to do this [here](https://github.com/fluxcd/terraform-provider-flux/tree/main/examples/helm-install)

**Comment by swade1987** [comment]
@bwestover 

Firstly, apologies for the slow response.

Please see the example [here](https://github.com/fluxcd/terraform-provider-flux/tree/main/examples/github-via-ssh). You can follow this but must remove the following block that creates the repository. 

```
# ==========================================
# Initialise a Github project
# ==========================================

resource "github_repository" "this" {
  name        = var.github_repository
  description = var.github_repository
  visibility  = "private"
  auto_init   = true # This is extremely important as flux_bootstrap_git will not work without a repository that has been initialised
}
```

You still need a token or SSH key with write access to Git.

**Comment by balonik** [comment]
@bwestover if you already have Git repository you can skip all the GitHub/GitLab/others related resources and just bootstrap Flux. The minimal code using `http` with GitLab could look like this.

```
provider "flux" {
  kubernetes = {
    # setup authentication here
  }
  git = {
    url                     = "https://gitlab.com/mygroup/myproject.git"
    http = {
      username = ...
      password = ...
    }
  }
}

resource "flux_bootstrap_git" "flux" {
  path                   = "<path to existing directory>"
}
```

**Comment by bwestover** [comment]
Thanks for the responses :bow: - I have this working now. 

To summarize the parts I was missing after reading the current docs and examples:
- **Whether or not this provider needs API access to the repo** (i.e. was the "github" provider in the examples incidental or is it a requirement)
  - In the GitHub examples creating a new repo used the "github" provider and token, but the Flux provider doesn't need it otherwise.
  - The provider handles the installation of Flux and configures it to talk to the repo. It does not handle the repo creation (contrasted with the Flux CLI which will create the repo if it doesn't exist). It can and will work with a pre-existing repo just fine.
- **Whether the path in the repo must exist already** 
  - The docs for the `path` variable say "... when specified the cluster sync will be scoped to this path", but it doesn't explicitly mention that it will be created if it doesn't exist. I assume it MUST create it since it works with brand new (empty) repositories)
- **How the authentication will be set up in the cluster** 
  - My frame of reference here is the Flux bootstrap CLI for GitHub (perhaps other integrations work differently). I've seen this work two ways:
    - If I pass both `--token-auth` and `--personal` flags I must provide a PAT, and that PAT gets stored in the "flux-system" secret and is used to make commits to the repo as part of the initial bootrapping.
    - If I pass only `--token-auth` then it is assumed that it is an organization owned repository. It uses the PAT to make commits to the repo during bootstrapping. It also creates a read-only deploy key, and stores that in the "flux-system" secret.

   TL;DR - In one case it will set up a deploy key and store it, and in the other case it just stores the PAT, however in BOTH cases it uses the PAT to make commits. What is the behavior of the Flux provider? So far it seems like however I set up auth in the Flux provider (in the nested git schema), is what gets stored on the cluster in the "flux-system" secret. If I want/need deploy keys, I need to set those up separately with Terraform (like in the [github-via-ssh](https://github.com/fluxcd/terraform-provider-flux/tree/main/examples/github-via-ssh) example.

FWIW, I don't disagree with any of these choices :smile_cat:  It seems like in Terraform we want separation of concerns vs the CLI which handles more parts of the process itself (e.g. creating deploy tokens, creating repo if necessary). I wanted to point out where my thinking was when trying to onboard to this provider in case it helps with documentation choices/efforts. Thanks for all you do here!

**Comment by swade1987** [comment]
@bwestover 

I love this, and thanks for the detailed reply. It's much appreciated.

I will definitely incorporate these into my next PR and "@" you to review it prior to merging.

**Comment by stefanprodan** [comment]
> If I pass only --token-auth then it is assumed that it is an organization owned repository. It uses the PAT to make commits to the repo during bootstrapping. It also creates a read-only deploy key, and stores that in the "flux-system" secret.

This is not the case, `flux bootstrap --token-auth` will create a secret with the token in the cluster and will configure Flux to clone over HTTP/S.

**Comment by NielsCodes** [comment]
I'm also finding the docs to be lacking. The examples seem mostly concerned with configuring the provider, but barely contain any content about the `flux_bootstrap_git` resource itself. For instance, which exact strings are supported for `components` and `components_extra` and why are these two separate properties to begin with? 

And even a simple example snippet right within the provider docs instead of only linking to Github examples would be a nice improvement imho.

**Comment by smheidrich** [comment]
**Update:** Complete nonsense comment hidden to not interrupt the discussion with it.

<details>

<summary>Old contents (wrong & based on incorrect understanding!)</summary>

One thing I didn't immediately understand, which perhaps the docs could make clearer: The general [Flux installation docs](https://fluxcd.io/flux/installation/) say:

>After running the bootstrap command, any operation on the cluster(s) (including Flux upgrades) can be done via Git push, without the need to connect to the Kubernetes API.

But when you use the Terraform provider, it overwrites the 3 files in the repo (`gotk-components.yaml`, `gotk-sync.yaml`, and `kustomize.yaml`) with each re-application of Terraform. Adding another file does nothing because `kustomization.yaml` is Flux's "entry point" that loads everything else, and files not referenced there will be ignored. So to actually use Flux as intended, usage of the [`kustomization_override`](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/resources/bootstrap_git#kustomization_override-2) field is _non-optional_: You **must** use this to specify extra YAML files that contain your own resources, otherwise there is no point using Flux and Terraform in the first place.

The provider docs don't really make this clear. They just contain 2 links with inconspicuous names like "Bootstrapping a cluster using a GitHub repository via SSH with flux customizations" in the middle of a long [list of examples](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/resources/bootstrap_git#example-usage). That isn't very prominent for something that is both _required_ and not really obvious.

I'm not the first person confused by this - here are some others who asked the exact same question:
- https://github.com/fluxcd/flux2/discussions/3716
- https://github.com/fluxcd/terraform-provider-flux/issues/478
- https://github.com/fluxcd/flux2/issues/5686

So in my opinion this should be made clearer. PR to try and do so:

- https://github.com/fluxcd/terraform-provider-flux/pull/820

</details>
